platforms/common/urlClient.cpp

```diff
@@ -17,186 +22,376 @@ struct CurlGlobals {
+
+    // This is probably not needed since all task have been canceled and joined
+    for (auto& task : m_tasks) {
+        if (task.active) {
+            curl_multi_remove_handle(m_curlHandle, task.handle);
+        }
+    }
+    curl_multi_cleanup(m_curlHandle);
+}
+
+void UrlClient::curlWakeUp() {
+
+    if (!m_curlNotified) {
+        if (write(m_requestNotify[1], "\0", 1) <= 0) {
+            // err
+            return;
+        }
+        //LOG("wake up!");
+        m_curlNotified = true;
     }
 }
 
-UrlClient::RequestId UrlClient::addRequest(const std::string& url, UrlCallback onComplete) {
-    // Create a new request.
+UrlClient::RequestId UrlClient::addRequest(const std::string& _url, UrlCallback _onComplete) {
+
     auto id = ++m_requestCount;
-    Request request = {url, onComplete, id, false};
+    Request request = {_url, _onComplete, id};
+
     // Add the request to our list.
     {
-        // Lock the mutex to prevent concurrent modification of the list by the curl loop thread.
+        // Lock the mutex to prevent concurrent modification of the
+        // list by the curl thread.
         std::lock_guard<std::mutex> lock(m_requestMutex);
         m_requests.push_back(request);
     }
-    // Notify a thread to start the transfer.
-    m_requestCondition.notify_one();
+    curlWakeUp();
+
     return id;
 }
 
-void UrlClient::cancelRequest(UrlClient::RequestId id) {
+void UrlClient::cancelRequest(UrlClient::RequestId _id) {
     UrlCallback callback;
     // First check the pending request list.
     {
-        // Lock the mutex to prevent concurrent modification of the list by the curl loop thread.
         std::lock_guard<std::mutex> lock(m_requestMutex);
         for (auto it = m_requests.begin(), end = m_requests.end(); it != end; ++it) {
             auto& request = *it;
-            if (request.id == id) {
-                // Found the request! Now run its callback and remove it.
+            if (request.id == _id) {
                 callback = std::move(request.callback);
                 m_requests.erase(it);
                 break;
             }
         }
     }
-    // We run the callback outside of the mutex lock to prevent deadlock in case the callback
-    // makes further calls into this UrlClient.
+
+    // We run the callback outside of the mutex lock to prevent deadlock
+    // in case the callback makes further calls into this UrlClient.
     if (callback) {
-        callback(getCanceledResponse());
+        UrlResponse response;
+        response.error = requestCancelledError;
+        callback(std::move(response));
+        return;
     }
 
-    // Next check the active request list.
-    for (auto& task : m_tasks) {
-        if (task.request.id == id) {
-            task.request.canceled = true;
+    // Check requests that are already active.
+    {
+        std::lock_guard<std::mutex> lock(m_requestMutex);
+        auto it = std::find_if(std::begin(m_tasks), std::end(m_tasks),
```

Comment:
nit: `m_tasks.begin()` seems simpler and clearer here.

Reply:
Somehow I find the free functions easier to parse - maybe it's just the syntax highlight  for `std::` that takes my focus to begin/end.